silkrpc/types/block.hpp

```diff
@@ -35,6 +37,70 @@ struct Block : public silkworm::BlockWithHash {
 
 std::ostream& operator<<(std::ostream& out, const Block& b);
 
+class BlockNumberOrHash {
+public:
+    BlockNumberOrHash() {}
+    BlockNumberOrHash(BlockNumberOrHash const& bnoh);
+    explicit BlockNumberOrHash(std::string const& bnoh) {
+        build(bnoh);
+    }
+    explicit BlockNumberOrHash(std::uint64_t const& number)
+        : number_{std::make_unique<std::uint64_t>(number)} {};
+
+    virtual ~BlockNumberOrHash() noexcept {}
+
+    BlockNumberOrHash& operator=(BlockNumberOrHash const& bnoh) = delete;
+    BlockNumberOrHash& operator=(std::string const& bnoh)  {
+        build(bnoh);
+        return *this;
+    }
+
+    BlockNumberOrHash& operator=(std::uint64_t const number) {
+        tag_.release();
+        hash_.release();
+        number_ = std::make_unique<std::uint64_t>(number);
+        return *this;
+    }
+
+    bool is_undefined() const {
+        return !(is_number() || is_tag() || is_hash());
+    }
+
+    bool is_number() const {
+        return number_.get() != nullptr;
+    }
+
+    uint64_t number() const {
+        return is_number() ? *number_ : 0;
+    }
+
+    bool is_hash() const {
+        return hash_.get() != nullptr;
+    }
+
+    evmc::bytes32 hash() const {
+        return is_hash() ? *hash_ : evmc::bytes32{0};
+    }
+
+    bool is_tag() const {
+        return tag_.get() != nullptr;
+    }
+
+    std::string tag() const {
+        return is_tag() ? *tag_ : "";
+    }
+
+private:
+    void build(std::string const& bnoh);
+    void set_number(std::string const& input, int base);
+
+    std::unique_ptr<std::uint64_t> number_;
+    std::unique_ptr<evmc::bytes32> hash_;
+    std::unique_ptr<std::string> tag_;
```

Comment:
I would really consider the usage of `std::variant` here: whenever possible, we should try to use types already provided in the standard library

Reply:
knowledge is power!! 
Done.